fuse_tutorials/include/fuse_tutorials/range_cost_functor.h

```diff
@@ -0,0 +1,135 @@
+ * form:
+ *   (Z - Z_hat)^T * S^{-1} * (Z - Z_hat)
+ * where S is the measurement covariance matrix.
+ */
+class RangeCostFunctor
+{
+public:
+  /**
+   * @brief Constructor
+   *
+   * The cost functor needs to know the details of the sensor measurement in order to compute costs for the optimizer.
+   * In the case of our range-only sensor, this includes the measured range to the landmark and the measurement
+   * uncertainty.
+   *
+   * @param[in] z The measured range to the landmark
+   * @param[in] sigma The standard deviation of the range measurement
+   */
+  RangeCostFunctor(const double z, const double sigma) : sigma_(sigma), z_(z) {}
+
+  /**
+   * @brief Compute the costs using the provided stored measurement details and the provided variable values.
+   *
+   * Ceres Solver will provide this functor with the current values of the robot position and landmark position.
+   * The functor must then populate the "residuals" vector with the computed costs. You can simply assume the size of
+   * all arrays are correct. See the costFunction() implementation of the RangeConstraint class for details on how the
+   * sizes and order of the variable are defined.
+   *
+   * Note the use of the template type T to describe the input and output arrays. This is a requirement when using
+   * automatic differentiation. Ceres will call the functor with "jet" datatypes sometimes, and with standard doubles
+   * at other time. This means that (a) we must implement our functor using the templated type, T. (b) Any constants
+   * that we use when computing the residuals/costs must be wrapped in a T object. And (c) we should use the
+   * Ceres Solver version of common math functions, such sin, cos, and sqrt, which have been optimized for use with
+   * Ceres Solver's jet datatypes. If you implement the Jacobian matrices manually, none of that is required.
+   *
+   * @param[in] robot_position - An array of component values (x, y) from the Position2DStamped variable representing
+   *                             the position of the robot in the global frame.
+   * @param[in] landmark_position - An array of component values (x, y) from the Point2DLandmark variable representing
+   *                                the position of the landmark in the global frame.
+   * @param[out] residuals - An array of computed cost values. In this case, our residual array is only a single value.
+   * @return True if the cost was computed successfully, false otherwise
+   */
+  template <typename T>
+  bool operator()(const T* const robot_position, const T* const landmark_position, T* residuals) const
+  {
+    // Implement our mathematic measurement model:
+    //   z_hat = sqrt( (x_robot - x_landmark)^2 + (y_robot - y_landmark)^2 )
+    //   error = (z - z_hat) / sigma
+    auto dx = robot_position[0] - landmark_position[0];
+    auto dy = robot_position[1] - landmark_position[1];
+    auto z_hat = ceres::sqrt(dx * dx + dy * dy);
```

Comment:
Why do I have it in my head that square roots in residual calculations are problematic? Or maybe I am thinking of inverse square roots, in which case the answer is obvious.

Reply:
Well, there is this:
http://ceres-solver.org/automatic_derivatives.html#pitfalls

I'd need to derive the actual derivatives, but I think Ceres's example still applies. So a robot driving through a beacon could potentially cause a problem. This could be fixed similar to here:
https://github.com/ceres-solver/ceres-solver/blob/master/include/ceres/rotation.h#L256-L274

```
auto norm_sq = dx * dx + dy * dy;
if (norm_sq > 0.0)
{
  auto z_hat = ceres::sqrt(norm_sq);
  residuals[0] = (T(z_) - z_hat) / T(sigma_);
}
else
{
  residuals[0] = T(z_) / T(sigma_);
}
return true;
```